### PR TITLE
 Single task execution (admin-generated workflow implementation)

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
This PR implements single task execution for CreateExecutionRequests with reference entities of type 'TASK'.

Because this is v1 of single task execution this code is compartmentalized and designed to be easy to rip out. Therefore not everything that could necessarily be refactored from the existing CreateExecution, CreateWorfklow and CreateLaunchPlan logic is changed and some duplicate code exists. This is (ugly) by design.

**TODO** If this overall design looks okay I will add more unit tests to this PR - I just wanted to get an overall approval of the code before investing time in unit tests in case it changes significantly.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
This code takes an existing, registered task and generates a skeleton workflow definition with a single node referencing the task. The workflow inputs and outputs are mapped to those of the task definition. For UI purposes, a default launch plan with a matching interface is also generated.

## Tracking Issue
https://github.com/lyft/flyte/issues/256

## Follow-up issue
https://github.com/lyft/flyte/issues/221

### CODE REVIEW

**Comment by wild-endeavor** [comment]
Is this really "v1".  I'd rather not call it that.

**Comment by katrogan** [comment]
@wild-endeavor updated but names are hard. let me know if you have suggestions

File: pkg/manager/impl/util/single_task_execution.go [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+	}
+	return bindings, nil
+}
+
+func CreateOrGetWorkflowModel(
+	ctx context.Context, request admin.ExecutionCreateRequest, db repositories.RepositoryInterface,
+	workflowManager interfaces.WorkflowInterface, namedEntityManager interfaces.NamedEntityInterface, taskIdentifier *core.Identifier,
+	task *admin.Task) (*models.Workflow, error) {
+	workflowModel, err := db.WorkflowRepo().Get(ctx, repositoryInterfaces.GetResourceInput{
+		Project: taskIdentifier.Project,
```
Comment (EngHabu) [inline pkg/manager/impl/util/single_task_execution.go:185]: We talked about creating these skeleton WFs in a dummy project/domain. So that we don't clutter the UI. Did that not work?

File: pkg/manager/impl/util/single_task_execution.go [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+		bindings = append(bindings, binding)
+	}
+	return bindings, nil
+}
+
+func CreateOrGetWorkflowModel(
+	ctx context.Context, request admin.ExecutionCreateRequest, db repositories.RepositoryInterface,
+	workflowManager interfaces.WorkflowInterface, namedEntityManager interfaces.NamedEntityInterface, taskIdentifier *core.Identifier,
+	task *admin.Task) (*models.Workflow, error) {
+	workflowModel, err := db.WorkflowRepo().Get(ctx, repositoryInterfaces.GetResourceInput{
```
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/manager/impl/util/single_task_execution.go:184]: true but i'm not sure how to guarantee against collisions using any naming scheme. couldn't a random hash or prefix/suffix pattern fall into the same trap? (i suppose with less likelihood)

File: pkg/manager/impl/util/single_task_execution.go (hunk 1 above)
Comment (katrogan) [reply pkg/manager/impl/util/single_task_execution.go:185]: unfortunately projects are registered dynamically although we could seed the database with a system project like we do in the sandbox but we also have https://github.com/lyft/flyteidl/blob/4c689bdb271db83c1934505282d5e72981f23a51/protos/flyteidl/admin/common.proto#L33 as a separate status denoting origin/status

File: pkg/manager/impl/util/single_task_execution.go [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+const maxNodeIDLength = 63
+
+var defaultRetryStrategy = core.RetryStrategy{
+	Retries: 3,
+}
+var defaultTimeout = ptypes.DurationProto(24 * time.Hour)
+
+func generateNodeNameFromTask(taskName string) string {
+	if len(taskName) >= maxNodeIDLength {
+		taskName = taskName[:maxNodeIDLength]
```
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/manager/impl/util/single_task_execution.go]: excellent point! updated

File: pkg/manager/impl/util/single_task_execution.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (EngHabu) [reply pkg/manager/impl/util/single_task_execution.go:184]: True, the entropy is much higher with random strings, but I guess introducing a random string means we will create this many many times (with every execution) creating a lot of garbage... 

Maybe just introducing a known constant obscure prefix/suffix is fine then?

File: pkg/manager/impl/util/single_task_execution.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+		bindings = append(bindings, binding)
+	}
+	return bindings, nil
+}
+
+func CreateOrGetWorkflowModel(
+	ctx context.Context, request admin.ExecutionCreateRequest, db repositories.RepositoryInterface,
+	workflowManager interfaces.WorkflowInterface, namedEntityManager interfaces.NamedEntityInterface, taskIdentifier *core.Identifier,
+	task *admin.Task) (*models.Workflow, error) {
+	workflowModel, err := db.WorkflowRepo().Get(ctx, repositoryInterfaces.GetResourceInput{
```
Comment (EngHabu) [inline pkg/manager/impl/util/single_task_execution.go:184]: Doesn't this introduce a race condition? if I tried to run the same task twice, both of them go past the "Get" call and realize there isn't an existing WF model, so they will both try to create one... Should the error check for Create() call below exclude "ErrExistingRow" or whatever the error is for that?

**Comment by katrogan** [comment]
> Are you also going to add the automatic filter for System Generated in the ListWF Endpoint?
Added the filter to the ListNamedEntity endpoint which is what the console uses to list workflows and is the entrypoint for individual workflows 

I'm more hesitant about adding it to the existing ListWorkflows endpoint by default because now that will force an additional join and I don't want to add a performance penalty to an endpoint that is really only available in the cli (aka power users who hopefully know what they're doing :) )

File: pkg/manager/impl/util/single_task_execution.go (hunk 4 above)
Comment (katrogan) [reply pkg/manager/impl/util/single_task_execution.go:198]: yes, good call. done

File: pkg/manager/impl/util/single_task_execution.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (katrogan) [reply pkg/manager/impl/util/single_task_execution.go:198]: sure done. let me know if you'd prefer a different name/format

**Comment by katrogan** [comment]
friendly ping @EngHabu

File: pkg/manager/impl/util/single_task_execution.go [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+			ResourceType: core.ResourceType_WORKFLOW,
+			Id: &admin.NamedEntityIdentifier{
+				Project: workflowIdentifier.Project,
+				Domain:  workflowIdentifier.Domain,
+				Name:    workflowIdentifier.Name,
+			},
+			Metadata: &admin.NamedEntityMetadata{State: admin.NamedEntityState_SYSTEM_GENERATED},
+		})
+		if err != nil {
+			logger.Warningf(ctx, "Failed to set skeleton workflow state to archived: %v", err)
```
Comment (EngHabu) [inline pkg/manager/impl/util/single_task_execution.go:270]: nit:
```suggestion
			logger.Warningf(ctx, "Failed to set skeleton workflow state to system-generated: %v", err)
```

**Comment by katrogan** [comment]
@EngHabu mind reviewing one last time? your suggested commit dismissed your review :)
